src/Firestore/ValueMapper.php

```diff
@@ -0,0 +1,499 @@
+     * @param mixed $value The firestore value.
+     * @return mixed
+     */
+    private function decodeValue($type, $value)
+    {
+        switch ($type) {
+            case 'booleanValue':
+            case 'nullValue':
+            case 'stringValue':
+            case 'doubleValue':
+                return $value;
+                break;
+
+            case 'bytesValue':
+                return new Blob($value);
+
+            case 'integerValue':
+                return $this->returnInt64AsObject
+                    ? new Int64($value)
+                    : (int) $value;
+
+            case 'timestampValue':
+                return $this->createTimestampWithNanos($value);
+                break;
+
+            case 'geoPointValue':
+                $value += [
+                    'latitude' => null,
+                    'longitude' => null
+                ];
+
+                return new GeoPoint($value['latitude'], $value['longitude']);
+                break;
+
+            case 'arrayValue':
+                $res = [];
+
+                foreach ($value['values'] as $val) {
+                    $type = array_keys($val)[0];
+
+                    $res[] = $this->decodeValue($type, current($val));
+                }
+
+                return $res;
+                break;
+
+            case 'mapValue':
+                $res = [];
+
+                foreach ($value['fields'] as $key => $val) {
+                    $type = array_keys($val)[0];
+
+                    $res[$key] = $this->decodeValue($type, current($val));
+                }
+
+                return $res;
+                break;
+
+            case 'referenceValue':
+                $parent = new CollectionReference($this->connection, $this, $this->parentPath($value));
+                return new DocumentReference($this->connection, $this, $parent, $value);
+
+            default:
+                throw new \RuntimeException(sprintf(
+                    'unexpected value type %s!',
+                    $type
+                ));
+
+                break;
+        }
+    }
+
+    /**
+     * Encode a Google Cloud PHP value as a Firestore value.
+     *
+     * @param mixed $value
+     * @return array [Value](https://firebase.google.com/docs/firestore/reference/rpc/google.firestore.v1beta1#value)
+     */
+    public function encodeValue($value)
+    {
+        $type = gettype($value);
+
+        switch ($type) {
+            case 'boolean':
+                return ['booleanValue' => $value];
+                break;
+
+            case 'integer':
+                return ['integerValue' => $value];
+                break;
+
+            case 'double':
+                return ['doubleValue' => $value];
+                break;
+
+            case 'string':
+                return ['stringValue' => $value];
+                break;
+
+            case 'resource':
+                return ['bytesValue' => base64_encode(stream_get_contents($value))];
```

Comment:
If I remember correctly, other gRPC APIs we've worked with don't require base64 encoding bytes. Is firestore unique?